File: 백준/Gold/2146. 다리 만들기/다리 만들기.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var (
	reader    = bufio.NewReader(os.Stdin)
	writer    = bufio.NewWriter(os.Stdout)
	visited   = [][]bool{}
	graph     = [][]int{}
	dx        = []int{-1, 1, 0, 0}
	dy        = []int{0, 0, -1, 1}
	islandNum = 1
	n         = 0
)

type Queue struct {
	X []int
	Y []int
}

func (q *Queue) Enqueue(x, y int) {
	q.X = append(q.X, x)
	q.Y = append(q.Y, y)
}

func (q *Queue) Dequeue() (x, y int) {
	if len(q.X) <= 0 {
		return -1, -1
	}
	x = q.X[0]
	y = q.Y[0]
	if len(q.X) == 1 {
		q.X = []int{}
		q.Y = []int{}
	} else {
		q.X = q.X[1:]
		q.Y = q.Y[1:]
	}
	return
}

func (q *Queue) Len() int {
	return len(q.X)
}

func devideIsland(i, j int) {
	island := new(Queue)
	island.Enqueue(i, j)
	var cx, cy int
	var nx, ny int
	for island.Len() > 0 {
		cx, cy = island.Dequeue()
		for i := 0; i < 4; i++ {
			nx = cx + dx[i]
			ny = cy + dy[i]
			if 0 <= nx && nx < n && 0 <= ny && ny < n && visited[nx][ny] == false && graph[nx][ny] == 1 {
				visited[nx][ny] = true
				graph[nx][ny] = islandNum
				island.Enqueue(nx, ny)
			}
		}
	}
}

func makingBridge(startIslandNum int) float64 {
	bridge := new(Queue)
	length := make([][]float64, n)
	for i := 0; i < n; i++ {
		length[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			length[i][j] = math.MaxFloat64
			if graph[i][j] == startIslandNum {
				length[i][j] = 0
				bridge.Enqueue(i, j)
			}
		}
	}

	var cx, cy int
	var nx, ny int
	for bridge.Len() > 0 {
		cx, cy = bridge.Dequeue()
		for i := 0; i < 4; i++ {
			nx = cx + dx[i]
			ny = cy + dy[i]
			if 0 <= nx && nx < n && 0 <= ny && ny < n {
				if graph[nx][ny] > 0 && graph[nx][ny] != startIslandNum && visited[nx][ny] {
					return length[cx][cy]
				} else if graph[nx][ny] == 0 && length[nx][ny] == math.MaxFloat64 {
					length[nx][ny] = length[cx][cy] + 1
					bridge.Enqueue(nx, ny)
				}
			}
		}
	}
	return math.MaxFloat64
}
func main() {
	defer writer.Flush()
	fmt.Fscanf(reader, "%d\n", &n)
	graph = make([][]int, n)
	visited = make([][]bool, n)
	var res float64
	res = math.MaxFloat64
	for i := 0; i < n; i++ {
		graph[i] = make([]int, n)
		visited[i] = make([]bool, n)
		for j := 0; j < n; j++ {
			fmt.Fscanf(reader, "%d ", &graph[i][j])
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if graph[i][j] == 1 && visited[i][j] == false {
				visited[i][j] = true
				graph[i][j] = islandNum
				devideIsland(i, j)
				islandNum++
			}
		}
	}
	for i := 1; i <= islandNum; i++ {
		res = math.Min(res, makingBridge(i))
	}
	fmt.Fprintf(writer, "%0.f\n", res)
}

File: 백준/Gold/2146. 다리 만들기/다리 만들기.py
from collections import deque
import sys
input = sys.stdin.readline
n = int(input())

graph = [list(map(int, input().split())) for _ in range(n)]
visited = [[False for _ in range(n)]for _ in range(n)]
dx = [-1,1,0,0]
dy = [0,0,-1,1]

def devide_island(i, j):
    island = deque()
    island.append([i, j])
    while island:
        cx, cy = island.popleft()
        for i in range(4):
            nx = cx+dx[i]
            ny = cy+dy[i]
            if 0<=nx<n and 0<=ny<n and visited[nx][ny]==False and graph[nx][ny]==1:
                visited[nx][ny] = True
                graph[nx][ny] = island_num
                island.append([nx,ny])

def making_bridge(start_island):
    q = deque()
    length = [[float('inf') for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if graph[i][j]==start_island:
                length[i][j] = 0
                q.append([i,j])
    while q:
        cx, cy = q.popleft()
        for i in range(4):
            nx = cx+dx[i]
            ny = cy+dy[i]
            if 0<=nx<n and 0<=ny<n:
                if graph[nx][ny]!=0 and graph[nx][ny]!=start_island and visited[nx][ny]:
                    return length[cx][cy]
                elif graph[nx][ny]==0 and length[nx][ny]==float('inf'):
                    length[nx][ny] = length[cx][cy]+1
                    q.append([nx,ny])
    return float('inf')
island_num=1
for i in range(n):
    for j in range(n):
        if graph[i][j]==1 and visited[i][j]==False:
            visited[i][j] = True
            graph[i][j] = island_num
            devide_island(i,j)
            island_num += 1

res = float('inf')
for i in range(1,island_num):
    res = min(res, making_bridge(i))
print(res)